clippy_lints/src/deref_coercions.rs

```diff
@@ -0,0 +1,55 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::adjustment::Adjust;
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for expressions that use deref coercion.
+    ///
+    /// ### Why is this bad?
+    /// Implicit deref coercion could result in confusing behavior when writing unsafe code.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x = &Box::new(true);
+    /// let y: &bool = x;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x = &Box::new(true);
+    /// let y: &bool = x.deref();
+    /// ```
+    #[clippy::version = "1.86.0"]
+    pub DEREF_COERCIONS,
+    pedantic,
```

Comment:
This should be a restriction lint since this restricts a language feature that isn't bad in itself and only makes sense to warn in specific situations